src/mlpack/core/metrics/bleu_impl.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file core/metrics/bleu_impl.hpp
+ * @author Mrityunjay Tripathi
+ *
+ * Implementation of BLEU class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_METRICS_BLEU_IMPL_HPP
+#define MLPACK_CORE_METRICS_BLEU_IMPL_HPP
+
+// In case it hasn't been included.
+#include "bleu.hpp"
+
+namespace mlpack {
+namespace metric {
+
+template <typename ElemType, typename PrecisionType>
+BLEU<ElemType, PrecisionType>::BLEU(const size_t maxOrder) :
+  maxOrder(maxOrder),
+  translationLength(0),
+  referenceLength(0)
+{
+  // Nothing to do here.
+}
+
+template <typename ElemType, typename PrecisionType>
+template <typename WordVector>
+std::map<WordVector, size_t> BLEU<ElemType, PrecisionType>::GetNGrams(
+  const WordVector& segment)
+{
+  std::map<WordVector, size_t> ngramsCount;
+  for (size_t order = 1; order < maxOrder + 1; ++order)
+  {
+    for (size_t i = 0; i < segment.size() - order + 1; ++i)
```

Comment:
Hmm... I think it's better to add an assert statement like --
```cpp
Log::Assert(segment.size() >= maxOrder);
```
or,
```cpp
if (segment.size() < maxOrder)
  Log::Fatal("Segment size cannot be smaller than 'maxOrder'.")
```
or,
```cpp
if (segment.size() < maxOrder)
{
  Log::Warn << "'maxOrder' greater than segment size. Setting 'maxOrder' to segment size...";
  maxOrder = segment.size();
}
```
What's your opinion?

Reply:
I think it's a normal situation. You don't need to raise an exception in this case. I guess something like this could work:
```c++
for (size_t order = 1; order < maxOrder + 1; ++order)
{
  for (size_t i = 0; i + order < segment.size() + 1; ++i)
```
Basically, it's the same check, but it avoids overflows.

Reply:
That makes sense. It took me sometime to understand the difference 😅